### DISCUSSION TITLE
Type Hinting for Structured Configs

### DISCUSSION THREAD

**Discussion by abefrandsen** [omry/omegaconf#1186] [Q&A] [answered]
Hi! I'm wondering if there is an established way to type hint a function argument that is a DictConfig but further adheres to a particular structured config schema. Let me illustrate two possibilities -- simply annotating with `DictConfig` (which is correct but non-specific), or annotating with the structured config schema (specific but incorrect):

```
from omegaconf import DictConfig
from dataclasses import dataclass

@dataclass
class FooConfig:
    name: str
    age: int

# type hint with DictConfig -- correct but nonspecific
def myfun(fooconfig: DictConfig):
    ...

# type hint with FooConfig -- incorrect but specific
def myfun(fooconfig: FooConfig):
    ...
```

Is there a way to annotate the `fooconfig` argument with a valid type hint that clearly indicates it should be consistent with the `FooConfig` structured config schema?

Thanks!

**Comment by abefrandsen** [comment]
So I think [this issue is relevant](https://github.com/omry/omegaconf/issues/899) -- looks like a generic `DictConfig` type could eventually be in the works. In my case above, it would look something like:
```
def myfun(fooconfig: DictConfig[FooConfig]):
    ...
```
In the meantime, is there any way apart from just using the plain `DictConfig` annotation?

**Comment by omry** [comment, marked as answer]
The recommended practice is the second approach (duck typing).
If you really need the underlying object type to match, you can convert the OmegaConf config to object.
See [OmegaConf.to_container](https://omegaconf.readthedocs.io/en/2.3_branch/usage.html#omegaconf-to-container) and [OmegaConf.to_object](https://omegaconf.readthedocs.io/en/2.3_branch/usage.html#omegaconf-to-object) (which is a syntactic sugar using to_container).
